pkg/internalapis/cnsvolumeoperationrequest/cnsvolumeoperationrequest.go

```diff
@@ -286,3 +290,87 @@ func (or *operationRequestStore) StoreRequestDetails(
+	log.Debugf("Deleting CnsVolumeOperationRequest instance with name %s/%s", csiconfig.DefaultCSINamespace, name)
+	err := or.k8sclient.Delete(ctx, &cnsvolumeoperationrequestv1alpha1.CnsVolumeOperationRequest{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: csiconfig.DefaultCSINamespace,
+		},
+	})
+	if err != nil {
+		if !apierrors.IsNotFound(err) {
+			log.Errorf("failed to delete CnsVolumeOperationRequest instance %s/%s with error: %v",
+				csiconfig.DefaultCSINamespace, name, err)
+			return err
+		}
+	}
+	return nil
+}
+
+// cleanupStaleInstances cleans up CnsVolumeOperationRequest instances for volumes that are no longer present in the
+// kubernetes cluster.
+func (or *operationRequestStore) cleanupStaleInstances(cleanupInterval int) {
+	ticker := time.NewTicker(time.Duration(cleanupInterval) * time.Minute)
+	ctx, log := logger.GetNewContextWithLogger()
+	log.Infof("CnsVolumeOperationRequest clean up interval is set to %d minutes", cleanupInterval)
+	for range ticker.C {
+		instanceMap := make(map[string]bool)
+
+		cnsVolumeOperationRequestList := &cnsvolumeoperationrequestv1alpha1.CnsVolumeOperationRequestList{}
+		err := or.k8sclient.List(ctx, cnsVolumeOperationRequestList)
+		if err != nil {
+			log.Errorf("failed to list CnsVolumeOperationRequests with error %v. Abandoning"+
+				"CnsVolumeOperationRequests clean up ...", err)
+			return
+		}
+
+		k8sclient, err := k8s.NewClient(ctx)
+		if err != nil {
+			log.Errorf("failed to get k8sclient with error: %v. Abandoning CnsVolumeOperationRequests"+
+				"clean up ...", err)
+			return
+		}
+		pvList, err := k8sclient.CoreV1().PersistentVolumes().List(ctx, metav1.ListOptions{})
+		if err != nil {
+			log.Errorf("failed to list PersistentVolumes with error %v. Abandoning"+
+				"CnsVolumeOperationRequests clean up ...", err)
+			return
+		}
+
+		for _, pv := range pvList.Items {
+			if pv.Spec.CSI != nil && pv.Spec.CSI.Driver == csitypes.Name {
+				instanceMap[pv.Name] = true
+				instanceMap[pv.Spec.CSI.VolumeHandle] = true
+			}
+		}
+
+		for _, instance := range cnsVolumeOperationRequestList.Items {
+			latestOperationDetailsLength := len(instance.Status.LatestOperationDetails)
+			if latestOperationDetailsLength != 0 &&
+				instance.Status.LatestOperationDetails[latestOperationDetailsLength-1].TaskStatus ==
+					TaskInvocationStatusInProgress {
+				continue
+			}
+			var trimmedName string
+			switch {
+			case strings.HasPrefix(instance.Name, "pvc"):
+				trimmedName = instance.Name
+			case strings.HasPrefix(instance.Name, "delete"):
+				trimmedName = strings.TrimPrefix(instance.Name, "delete-")
+			case strings.HasPrefix(instance.Name, "extend"):
+				trimmedName = strings.TrimPrefix(instance.Name, "extend-")
+			}
+			if _, ok := instanceMap[trimmedName]; !ok {
+				err = or.deleteRequestDetails(ctx, instance.Name)
+				if err != nil {
+					log.Errorf("failed to delete CnsVolumeOperationRequest instance %s with error %v",
+						instance.Name, err)
```

Comment:
Do we need to update some sort of status for the instance when delete fails?

Reply:
Nope. There's no controller watching on these objects. We just attempt to delete and if it fails, try again in thenext cycle